homeassistant/components/teslemetry/entity.py

```diff
@@ -1,52 +1,114 @@
 )
 from .models import TeslemetryEnergyData, TeslemetryVehicleData
 
 
-class TeslemetryVehicleEntity(CoordinatorEntity[TeslemetryVehicleDataCoordinator]):
-    """Parent class for Teslemetry Vehicle Entities."""
+class TeslemetryEntity(
+    CoordinatorEntity[
+        TeslemetryVehicleDataCoordinator | TeslemetryEnergySiteLiveCoordinator
+    ]
+):
+    """Parent class for all Teslemetry entities."""
 
     _attr_has_entity_name = True
 
     def __init__(
         self,
-        vehicle: TeslemetryVehicleData,
+        coordinator: TeslemetryVehicleDataCoordinator
+        | TeslemetryEnergySiteLiveCoordinator,
+        api: VehicleSpecific | EnergySpecific,
         key: str,
     ) -> None:
         """Initialize common aspects of a Teslemetry entity."""
-        super().__init__(vehicle.coordinator)
+        super().__init__(coordinator)
+        self.api = api
         self.key = key
-        self.api = vehicle.api
-        self._wakelock = vehicle.wakelock
+        self._attr_translation_key = self.key
+        self._async_update_attrs()
 
-        car_type = self.coordinator.data["vehicle_config_car_type"]
+    @property
+    def available(self) -> bool:
+        """Return if sensor is available."""
+        return self.coordinator.last_update_success and self._attr_available
 
-        self._attr_translation_key = key
-        self._attr_unique_id = f"{vehicle.vin}-{key}"
-        self._attr_device_info = DeviceInfo(
-            identifiers={(DOMAIN, vehicle.vin)},
-            manufacturer="Tesla",
-            configuration_url="https://teslemetry.com/console",
-            name=self.coordinator.data["vehicle_state_vehicle_name"],
-            model=MODELS.get(car_type, car_type),
-            sw_version=self.coordinator.data["vehicle_state_car_version"].split(" ")[0],
-            hw_version=self.coordinator.data["vehicle_config_driver_assist"],
-            serial_number=vehicle.vin,
-        )
+    @property
+    def _value(self) -> Any | None:
+        """Return a specific value from coordinator data."""
+        return self.coordinator.data.get(self.key)
+
+    def get(self, key: str, default: Any | None = None) -> Any | None:
+        """Return a specific value from coordinator data."""
+        return self.coordinator.data.get(key, default)
+
+    @property
+    def is_none(self) -> bool:
+        """Return if the value is a literal None."""
+        return self.get(self.key, False) is None
+
+    def has(self, key: str | None = None) -> bool:
+        """Return True if a specific value is in coordinator data."""
+        return (key or self.key) in self.coordinator.data
+
+    async def handle_command(self, command) -> dict[str, Any]:
+        """Handle a command."""
+        try:
+            result = await command
+            LOGGER.debug("Command result: %s", result)
+        except TeslaFleetError as e:
+            LOGGER.debug("Command error: %s", e.message)
+            raise ServiceValidationError(
```

Comment:
Is this really a validation error or simply the service returning with an error.
We should separate a validation problem ( `ServiceValidationError`) or other issues such as connectivity etc. (`HomeAssistantError`)

Reply:
Seems to be more cases also below

Reply:
I recall being told or reading that service failures were `ServiceValidationError` not `HomeAssistantError`, but I'm probably wrong. I have changed them all over.

Reply:
https://developers.home-assistant.io/blog/2023/11/30/service-exceptions-and-translations#exception-handling-during-service-calls